cli/clibase/net.go

```diff
@@ -0,0 +1,50 @@
+package clibase
+
+import (
+	"net"
+	"strconv"
+
+	"github.com/pion/udp"
+	"golang.org/x/xerrors"
+)
+
+// Net abstracts CLI commands interacting with the operating system networking.
+//
+// At present, it covers opening local listening sockets, since doing this
+// in testing is a challenge without flakes, since it's hard to pick a port we
+// know a priori will be free.
+type Net interface {
+	// Listen has the same semantics as `net.Listen` but also supports `udp`
+	Listen(network, address string) (net.Listener, error)
+}
+
+// osNet is an implementation that call the real OS for networking.
+type osNet struct{}
+
+func (osNet) Listen(network, address string) (net.Listener, error) {
```

Comment:
How would we handle situations where you might need to configure dialing via `net.Dialer`?

Maybe we don't have such use-cases at this time but I'm concerned since we are introducing the field on clibase, which suggests it's the only way one should establish network connections (compared to injecting it into the specific command that needs it).

Reply:
Add a `GetDialer(options) net.Dialer` method to the `Net` interface, or somesuch.

> (compared to injecting it into the specific command that needs it)

I'm not sure what you mean here.  How would that suggestion be different than what I've implemented?

Reply:
The idea here is that `Net` is like the stdio and environment fields on `Invocation`.  When you actually run the command from a command line, you get the OS-provided functions, but in testing we can hook various things up to intercept the command interacting with the OS, just like how today we hook stdio into the tests.

Reply:
> Add a `GetDialer(options) net.Dialer` method to the `Net` interface, or somesuch.

I suppose that could work, injecting resolver/dialer for the in-mem stuff 👍🏻.

> I'm not sure what you mean here. How would that suggestion be different than what I've implemented?

With that parenthesis, I basically meant not putting the field in clibase/inv until there are more use-cases, i.e. keeping it port-forwarding only for now.

Reply:
I'm not sure how we could keep it port-forwarding only without a major refactor. The command handler only gets an `Invocation`, which is also the only thing we have access to from `clitest.New()`. There literally isn't anything else to hang it on besides `Invocation` (module-level `var`s will wreak havoc in parallel testing).

Reply:
There's also `context.Context`, but I hear you. That's perhaps a bit hidden away.